src/AbstractHttpMessageProducer.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * This file is part of the prooph/psb-http-producer.
+ * (c) 2014-2017 prooph software GmbH <contact@prooph.de>
+ * (c) 2015-2017 Sascha-Oliver Prolic <saschaprolic@googlemail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Prooph\ServiceBus\Message\Http;
+
+use Http\Discovery\MessageFactoryDiscovery;
+use Http\Message\RequestFactory;
+use Prooph\Common\Messaging\Message;
+use Prooph\Common\Messaging\MessageConverter;
+use Prooph\Common\Messaging\MessageDataAssertion;
+use Prooph\ServiceBus\Async\MessageProducer;
+use Prooph\ServiceBus\Exception\RuntimeException;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\UriInterface;
+use React\Promise\Deferred;
+
+abstract class AbstractHttpMessageProducer implements MessageProducer
+{
+    /**
+     * @var RequestFactory
+     */
+    protected $requestFactory;
+
+    /**
+     * @var MessageConverter
+     */
+    protected $messageConverter;
+
+    /**
+     * @var UriInterface
+     */
+    protected $uri;
+
+    public function __construct(
+        MessageConverter $messageConverter,
+        UriInterface $uri,
+        RequestFactory $requestFactory = null
+    ) {
+        $this->messageConverter = $messageConverter;
+        $this->uri = $uri;
+        $this->requestFactory = $requestFactory ?: MessageFactoryDiscovery::find();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __invoke(Message $message, Deferred $deferred = null): void
+    {
+        $messageData = $this->messageConverter->convertToArray($message);
+
+        MessageDataAssertion::assert($messageData);
+
+        $messageData['created_at'] = $message->createdAt()->format('Y-m-d\TH:i:s.u');
+
+        $request = $this->requestFactory->createRequest(
+            'POST',
+            $this->uri,
+            [
+                'Content-Type' => 'application/json'
+            ],
+            json_encode($messageData)
+        );
+
+        $this->handleRequest($request, $deferred);
+    }
+
+    /**
+     * @param ResponseInterface $response
+     * @return array|mixed
+     * @throws RuntimeException
+     */
+    protected function getPayloadFromResponse(ResponseInterface $response)
+    {
+        $payload = json_decode($response->getBody(), true);
```

Comment:
I think it might be time for a `\Prooph\Common\json_decode` function I've seen this in quite a few places.